godot-core/src/builtin/rid.rs

```diff
@@ -0,0 +1,140 @@
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+use std::num::NonZeroU64;
+
+use godot_ffi as sys;
+use sys::{ffi_methods, GodotFfi};
+
+/// A RID ("resource ID") is an opaque handle that refers to a Godot `Resource`.
+///
+/// RIDs do not grant access to the resource itself. Instead, they can be used in lower-level resource APIs
+/// such as the [servers]. See also [Godot API docs for `RID`][docs].
+///
+/// RIDs should be largely safe to work with. Certain calls to servers may fail, however doing so will
+/// trigger an error from Godot, and will not cause any UB.
+///
+/// # Layout:
+///
+/// `Rid` has the same layout as a [`u64`], where [`Rid::Invalid`] is 0, and [`Rid::Valid(i)`] is `i`.
+///
+/// # Safety Caveat:
+///
+/// In Godot 3, RID was not as safe as described here. We believe that this is fixed in Godot 4, but this has
+/// not been extensively tested as of yet. Some confirmed UB from Godot 3 does not occur anymore, but if you
+/// find anything suspicious or outright UB please open an issue.
+///
+/// [servers]: https://docs.godotengine.org/en/stable/tutorials/optimization/using_servers.html
+/// [docs]: https://docs.godotengine.org/en/stable/classes/class_rid.html
+
+// Using normal rust repr to take advantage advantage of the nullable pointer optimization. As this enum is
+// eligible for it, it is also guaranteed to have it. Meaning the layout of this type is identical to `u64`.
+// See: https://doc.rust-lang.org/nomicon/ffi.html#the-nullable-pointer-optimization
+// Cannot use `#[repr(C)]` as it does not use the nullable pointer optimization.
+#[derive(Copy, Clone, Eq, PartialEq, Hash, Ord, PartialOrd, Debug)]
+pub enum Rid {
+    /// A valid RID may refer to some resource, but is not guaranteed to do so.
+    Valid(NonZeroU64),
+    /// An invalid RID will never refer to a resource. Internally it is represented as a 0.
+    Invalid,
+}
+
+impl Rid {
+    /// Create a new RID.
+    #[inline]
+    pub const fn new(id: u64) -> Self {
+        match NonZeroU64::new(id) {
+            Some(id) => Self::Valid(id),
+            None => Self::Invalid,
+        }
+    }
+
+    /// Convert this RID into a [`u64`].
+    ///
+    /// _Godot equivalent: `Rid.get_id()`_
+    #[inline]
+    pub const fn to_u64(self) -> u64 {
+        match self {
+            Rid::Valid(id) => id.get(),
+            Rid::Invalid => 0,
+        }
+    }
+
+    /// Convert this RID into a [`NonZeroU64`].
+    #[inline]
+    pub const fn to_non_zero_u64(self) -> Option<NonZeroU64> {
+        match self {
+            Rid::Valid(id) => Some(id),
+            Rid::Invalid => None,
+        }
+    }
+
+    /// Returns `true` if this is a valid RID.
+    #[inline]
+    pub const fn is_valid(&self) -> bool {
+        matches!(self, Rid::Valid(_))
+    }
+
+    /// Returns `true` if this is an invalid RID.
+    #[inline]
+    pub const fn is_invalid(&self) -> bool {
+        matches!(self, Rid::Invalid)
+    }
```

Comment:
Providing negated versions of accessors is not something we typically do. I don't think we should start a precedent here, as this will lead to discussions in which cases it's OK or not. Given that RIDs validity checks should ideally not be very common, it's fine if a user has to explicitly type a `!`.

Reply:
oh fair, i think the c++ code had it so i just added it. But it does follow the precedent of `Option` which has both `is_some` and `is_none`.

Reply:
> But it does follow the precedent of Option which has both is_some and is_none.

Yes, and it does _not_ follow the precedent of `Vec::is_empty()` which has no such opposite. Or any other type in godot-rust, which is much more relevant here 🙂

Reply:
`Vec` isn't an enum though, i chose to compare it with `Option` because they're both similar enums, and having an `is_x` function for each enum variant is not uncommon.